tools/symbols_map_generator.py

```diff
@@ -0,0 +1,343 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+import glob
+import os
+import argparse
+from pathlib import Path
+from typing import Literal, TypedDict, get_args, Optional
+import clang.cindex
+
+_logger = logging.getLogger(__name__)
+
+LibraryName = Literal[
+    "miral", "mir_common", "mir_core", "miroil", "mir_platform",
+    "mir_renderer", "mir_renderers", "mir_server", "mir_test", "mir_wayland"]
+
+
+class LibraryInfo(TypedDict):
+    external_headers: list[str]
+    internal_headers: Optional[list[str]]
+    external_output_file: str
+    internal_output_file: Optional[str]
+    search_headers: Optional[list[str]]
+
+
+# Directory paths are relative to the root of the Mir project
+library_to_data_map: dict[LibraryName, LibraryInfo] = {
+    "miral": {
+        "external_headers": [
+            "include/miral"
+        ],
+        "external_output_file": "src/miral/symbols.map",
+    },
+    "mir_server": {
+        "external_headers": [
+            "include/server/mir",
+            "src/include/server/mir"
+        ],
+        "external_output_file": "src/server/symbols.map",
+        "search_headers": [
+            "src/include/server"
+        ]
+    }
+}
+
+
+def get_absolute_path_from_project_path(project_path: str) -> Path:
+    return Path(__file__).parent.parent / project_path
+
+
+def is_operator_overload(spelling: str):
+    operators = [
+        "+",
+        "-",
+        "*",
+        "/",
+        "%",
+        "^",
+        "&",
+        "|",
+        "~",
+        "!",
+        "=",
+        "<",
+        ">",
+        "+=",
+        "-=",
+        '*=',
+        "/=",
+        "%=",
+        "^=",
+        "&=",
+        "|=",
+        "<<",
+        ">>",
+        ">>=",
+        "<<=",
+        "==",
+        "!=",
+        "<=",
+        ">=", 
+        "<=>",
+        "&&",
+        "||",
+        "++",
+        "--",
+        ",",
+        "->*",
+        "->",
+        "()",
+        "[]"
+    ]
+
+    if spelling.startswith("operator "):
+        return True
+
+    for op in operators:
+        if spelling == f"operator{op}":
+            return True
+        
+    return False
+
+
+def create_symbol_name(node: clang.cindex.Cursor) -> str:
+    if(node.kind == clang.cindex.CursorKind.CLASS_DECL
+        or node.kind == clang.cindex.CursorKind.STRUCT_DECL
+        or node.kind == clang.cindex.CursorKind.FIELD_DECL):
+        return node.displayname
+    
+    if node.kind == clang.cindex.CursorKind.DESTRUCTOR:
+        return f"?{node.spelling[1:]}*"
+    
+    if ((node.kind == clang.cindex.CursorKind.FUNCTION_DECL 
+        or node.kind == clang.cindex.CursorKind.CXX_METHOD
+        or node.kind == clang.cindex.CursorKind.CONVERSION_FUNCTION)
+        and is_operator_overload(node.spelling)):
+        return "operator*"
```

Comment:
So I got non-globbing working _somewhat_, but it seems like more of a rabbit hole than I'd like. For example, clang will parse `void do_x(int const& y)` to `void do_x(const int & y)`. These are the same thing, but if I output the symbol as the latter, then the former won't resolve. It seems to be an exact match, rather than a semantic match? I suppose I could go and find the exact line in the header file and copy and paste that somehow though... I will check that out. It' super finicky though

Reply:
I don't get what you have tried and what didn't work.

The actual symbol that gets matched by the linker is the mangled form (something like `_ZN4do_xERKi` in your example). I don't see how the left const or right const spelling can affect that. All we need to do is be consistent in the representation used in the symbol file.

Reply:
Somehow, the west vs east const spelling does affect it. I find this odd too, but when the `const` matches that found in the file, then the symbols resolves. Otherwise, it fails to resolve. I will confirm this with an example, but it is odd to me as well.

Reply:
Maybe I'm doing something wrong? As an example:

```diff
- miral::ExternalClientLauncher::launch*;
+ "miral::ExternalClientLauncher::launch(std::vector<std::string> const& command_line)";
```

This doesn't work

Reply:
Also, I _can_ output the mangled name here. Could we just do that?